docs/csharp/roslyn-sdk/incremental-source-generators/overview.md

```diff
@@ -0,0 +1,158 @@
+
+Source generators run during both design time builds and [[full]] builds. This allows them to extend user code with new code that adds classes, methods and properties that immediately affect the users experience via features like IntelliSense. It also means that generators run very often, possibly on every keystroke.
+
+Most programmers will not need to write source generators, but will take advantage of those written by others. If you do write a generator, follow the [Performance Guidelines]() and be [tested for performance]() because your generator can adversely affect the design time performance of editors such as Visual Studio. Source generators are new and we are evolving guidance on banned APIs, so feel free to reach out at [[@Jaredpa and @chsienki where?]] if you want to check the impact of a technique.
+
+> [!CAUTION]
+> We will disable generators during the design time build if they are slow or we analysis indicates that they do not follow guidelines.
+
+This article assumes that you understand the basics of the [Roslyn compiler](), [syntax trees](), [compilations](), [compilation](), [cancellation tokens]() and the [difference between value and reference equality](). It will also help to have an understanding of functional techniques using delegates such as lambda expressions.
+
+## How Roslyn generators work
+
+When the Roslyn compiler runs without generation, it reads user code and builds a set of syntax trees, generally one syntax tree per file. Based on this, the compiler builds a compilation and emits [IL]() based on the compilation:
+
+```mermaid
+flowchart LR
+    SyntaxTrees-->Compilation-->IL
+```
+
+When generators are present, the Roslyn compiler inserts an extra step between creating the compilation and emitting the IL which runs the generators and adds any syntax trees they create. It then rebuilds the compilation before emitting the IL:
+
+```mermaid
+flowchart LR
+    SyntaxTrees-->Comp1[Compilation]-->Generators-->More[More SyntaxTrees]-->Comp2[Compilation]-->IL
+    SyntaxTrees-->Generators
+    SyntaxTrees-->Comp2
+```
+
+Syntax trees are a full representation of the code the user wrote. The compilation includes several things that may be interersting to your generator, including the semantic model.
+
+In some cases additional information is needed, perhaps from a configuration file:
+
+```mermaid
+flowchart LR
+    ConfigFile[Config file]-->Generators
+    SyntaxTrees-->Comp1[Compilation]-->Generators-->More[More SyntaxTrees]-->Comp2[Compilation]-->IL
+    SyntaxTrees-->Generators
+    SyntaxTrees-->Comp2
+```
+
+In some cases, such as creating a strongly typed wrapper for a comma delimited file (`.csv`), the initial compilation might not be used by the generator, but it is always available to it.
+
+Each generator uses the same set of syntax trees and compilation, so generators are independent and do not know what was produced by other generators.
+
+```mermaid
```

Comment:
Again, mermaid live could be used here. For example, consider this SVG.

https://mermaid.ink/svg/pako:eNqFks1OwzAQhF_F2nMrFOeWAxIChJAKB8rN4bDE2yYitiPbUUFV351NS6iIiboHy5oZfbv-2UPlNEEBm9btqhp9FKuX0gqu0L9vPXa12JIlj9H5oMTbyRuK5Uw9jN5fQ84ZeWKQ1ZN-4ctG_Lyx-taZrmkxNs6q9VG8qs7SL_iUf_VEYZSGWDbl_4NdLq8nx2DFOE-ZeuJVrM_owRmwUj3TTqRzzONlipfz-Eu0PKXll2jH60glydLjChZgyBtsNH-D_eCWEGsyVELBW43-o4TSHjjXdxoj3euGm0OxwTbQArCPjltXUETf0xi6a5Bf0_ykDt9KvNIc